### ISSUE TITLE
For doesn't create identical elements

### ISSUE THREAD

**Issue by bluenote10** [ryansolid/solid#50]
I noticed that `For` doesn't want to create the same repeated element multiple times. The following example ([codesandbox](https://codesandbox.io/s/debugfor-6h0bc)) should output multple divs all showing "new element", but only one element gets render:

```js
import { createState, createEffect } from "solid-js";
import { render, For } from "solid-js/dom";

const App = () => {
  const [state, setState] = createState({
    list: []
  });

  setInterval(() => {
    setState({
      list: state.list.concat(["new element"])
    });
  }, 1000);

  createEffect(() => {
    console.log(state.list);
  });

  /* prettier-ignore */
  return (
    <For each={(state.list)}>
      {(item) => <div>{item}</div>}
    </For>
  );
};

render(App, document.getElementById("app"));
```

If I give the elements unique values it works.

**Comment by ryansolid** [comment]
Ok I will give it a look, I admittedly am not terribly surprised given the way it value checks to decide the "key". So I assume this was happening before, but I will take a look. The old version was something I had Frankenstein'd, whereas the new version is actually almost a verbatim port of Surplus and S.js.

**Comment by ryansolid** [comment]
Ok, I figured it out. Turns out an optimization I was doing actual avoided a code path for these cases. I haven't hit any issues with sorting etc, but I have to imagine that there are some holes on duplicate listings on simple string values (and there always has been). I'm adding back the optimization which incidentally fixes this case, I'm just concerned there may be others.

Released in v0.9.2

**Comment by bluenote10** [comment]
:+1: Not sure if the only other issue I've noticed so far is related: It happens when using a mapping function of the form `(s: string, i: number) => <div>{"is activated: " + (i == state.activeIndex)}</div>`. But I still have to reproduce that in a minimal test case.
